### ISSUE TITLE
read_next_chunk example unconvincing

### ISSUE THREAD

**Issue by shaneasd** [hsutter/708#25]
I'm not totally convinced that out parameters need to exist at all (i.e. #5) but this issue isn't about that. It's specifically about the argument that they make "caller-allocated out." better. I recognize that in https://github.com/hsutter/708/issues/5#issuecomment-735310044 it is pointed out that this is the less important reason for the existence of out parameters but if the argument is flawed (which I'll attempt to convince you that it is) then it would be better to remove it from the paper to keep focus on the major justification.

Firstly my understanding is that a out parameter must either directly construct or be assigned to (depending on whether it's uninitialized). That is to say that when this function is executed
```
void read_next_chunk(out db_format x)
{
    x = db_format(0);
}
```
it will either execute the custom constructor accepting an int directly into the argument (not unlike copy elision) or it will custom construct a db_format on the stack and then assign (move assign if possible) into the argument. This is based on the comment in ```int f()``` on page 22 in 2.1. It also appears to be supported by the work in progress reference implementation ( https://cppx.godbolt.org/z/EEWjd6 ) If this understanding is incorrect you can probably disregard everything that follows.

Fundamentally my concern is that in the latter case there is a construction and move that shouldn't be happening. In fact this is the exact criticism leveled in the paper against the return-by-value alternative. It seems to me that in general the goal is to avoid constructing ```db_format``` as much as possible and as such the implementation of ```read_next_chunk``` I've provided is actually pretty unlikely. Far more likely that ```read_next_chunk``` would mutate ```x``` without actually constructing a new one (e.g. using a memcpy).

In fact the paper identifies this exact use case in the case of read_next_chunk which accepts a vector<byte>. The assertion is that this function would mutate the function using ```data``` an similar functions. The paper asserts that this would work the same way but if my understanding is correct then I don't see how it could due to the new  requirement to assign to (as distinct from merely mutating) the out parameter.

Speculating as to where the issue lies:
- If the intended implementation is as above and there is a construct+move every call to read_next_chunk after the first then I think the return-by-value alternative is marginally superior in terms of expressiveness and pass by inout would be superior in terms of avoiding unnecessary operations.
- If the intended implementation is as above and there is not a move but rather x is reconstructed in place every time then that still seems like more construction than should be necessary.
- If the intended implementation is anything less than a complete reassignment of the object then it seems like it would be very difficult to say whether the out parameter had or had not in fact been supplied to the caller. It would be very easy to violate the rule that the previous value of the argument should have no direct impact on its new value.

**Comment by hsutter** [comment]
Good point, I can clarify this better in the paper.

> void read_next_chunk(out db_format x)
> {
>     x = db_format(0);
> }
> it will either execute the custom constructor accepting an int directly into the argument (not unlike copy elision) or it will custom construct a db_format on the stack and then assign (move assign if possible) into the argument.

The intent is that for an `out` parameter whose argument is already initialized, it's still actually an assignment. I'll expand 2.1 bullet 2 as follows:

   ... and that definite first use must be either to pass `x` to another function’s `out` parameter, or else be an assignment-expression to `x` (which can be from a single-object expression or a possibly-empty `initializer_list`) that performs a single constructor call (usually direct-initialization) of `x` with the right-hand side expression treated as the argument to `x`’s constructor __if the variable is uninitialized (this will always be true for uninitialized local variables, and true for `out` parameters when passed an uninitialized argument), otherwise performs an assignment__.

There's probably a better way to break up that sentence but it's good enough to capture this point for now.

> It also appears to be supported by the work in progress reference implementation

Note that `out` parameters are not yet implemented, so I wouldn't draw any conclusions from the prototype about `out` yet.

Thanks!

**Comment by hsutter** [comment]
On second thought, I'm not going to make this change because this section is only talking about local variables, not `out` parameter. That this works as I described (using assignment for already-constructed variables) is already covered in the 1.1.2 table for `out` parameter:

> If the first use is an assignment-expression, then if the argument is uninitialized it constructs `x` with the right-hand side expression treated as the constructor argument list; otherwise, it performs an assignment as usual

**Comment by shaneasd** [comment]
From what you've said it sounds like my understanding _is_ correct and as such no clarification is needed. However that leaves my actual concern (everything after "Fundamentally my concern is") unaddressed.

Presumably most of the time you're doing an assignment such as this you'd be assigning from a newly constructed object of the same type. Doesn't that mean each repeated call to read_next_chunk will involve an unnecessary construct+assign?

**Comment by hsutter** [comment]
> Doesn't that mean each repeated call to read_next_chunk will involve an unnecessary construct+assign?

I don't think so, and when it does it can be very cheap `x = {};`.

Longer answer:

**Background:** The canonical example is to reuse the capacity of a `string` or `vector` when calling `get_next_chunk` in loop. If we use an `out` parameter (today, a `&` parameter) we only need to reallocate if the capacity isn't already big enough, and so typically only a few loop iterations will allocate memory and then there are no allocations at all once we get to the biggest chunk for that loop (if the chunks are all the same size, we allocate only once for the whole loop). That's a big performance savings compared to returning by value which forces an allocation on every call.

**When you don't need to construct a temporary object:** Inside `get_next_chunk`, the loop body may already have a suitable string or character buffer it can just assign from. For example, `string` can call `operator=` to assign from anything we can view with a `string_view` and so `operator=` copy the contents of any contiguous character buffer and the function can execute without any object construction or reallocation if the capacity is already big enough.

**When you do construct a temporary but it's cheap:** In general, even if the type doesn't have a suitable `operator=`, you can usually use the technique of `x = {};` (default assignment, and yes that's 'from a newly constructed object' but it's usually super cheap and for types like `string` and `vector` won't change capacity) followed by using any non-`const` functions to set specific values in `x` and you still have the nice guarantee on exit that all elements in the `string` or `vector` really are ones that the function wrote to (avoiding the usual problems with fill-array APIs related to 'well but how many elements did you actually write to and how many did you leave uninitialized' -- not a problem with this technique).

I mention "default assignment" in section 2.3, and to the final example in 2.3 I should add:

> Note this already works for some types, including vector, using {} syntax:
> `v = {};					// default assignment (already legal)`


I should probably add this as a note though! Will do.

**Comment by shaneasd** [comment]
Good point. There's a key difference between returning a large struct (where return by value seems fine to me) and returning a small but expensive to construct struct (like vector). I had overlooked the fact that v = {} will disambiguate to an empty initializer list rather than a default constructed vector. Thanks for clarifying.
